VebTree.cs

```diff
@@ -82,8 +82,335 @@ public interface IVebTree
+                if (hasChildren[siblingId] || exists[siblingId]) { break; }
+                hasChildren[parentId] = false;
+            }
+        }
+
+        private ulong? getMinChild(ulong id, int rank)
+        {
+            ulong min = id;
+
+            // loop until there are no children to be explored
+            while (hasChildren[min])
+            {
+                // visit the child on the left side (if exists)
+                if (hasLeftChildren(min, rank)) { min = getLeftChild(min, rank--); }
+
+                // check if the visited node is the min. child -> terminate
+                else if (exists[min]) { break; }
+
+                // no lefthand children, parent and its other children are smaller
+                // -> explore righthand children (there have to be nodes)
+                else { min = getRightChild(min, rank--); }
+            }
+
+            // return the greatest child or null if there is none
+            return exists[min] ? min : null;
+        }
+
+        private ulong? getMaxChild(ulong id, int rank)
+        {
+            ulong max = id;
+
+            // loop until there are no children to be explored
+            while (hasChildren[max])
+            {
+                // visit the child on the right side (if exists)
+                if (hasRightChildren(max, rank)) { max = getRightChild(max, rank--); }
+
+                // check if the visited node is the max. child -> terminate
+                else if (exists[max]) { break; }
+
+                // no righthand children, parent and its other children are greater
+                // -> explore lefthand children (there have to be nodes)
+                else { max = getLeftChild(max, rank--); }
+            }
+
+            // return the greatest child or null if there is none
+            return exists[max] ? max : null;
+        }
+
+        private bool hasLeftChildren(ulong id, int rank)
+        {
+            if (rank == 0) { return false; }
+            ulong leftChild = this.getLeftChild(id, rank);
+            return hasChildren[leftChild] || exists[leftChild];
+        }
+
+        private bool hasRightChildren(ulong id, int rank)
+        {
+            if (rank == 0) { return false; }
+            ulong rightChild = this.getRightChild(id, rank);
+            return hasChildren[rightChild] || exists[rightChild];
+        }
+
+        private int getNodeRank(ulong id)
+        {
+            // this is based on a property of the binary tree nodes where ids
+            // consisting of 2^r (for max. r) are positioned at tree rank r,
+            // e.g. odd ids are positioned at the lowest rank with r=0, ids
+            // divisible by 2^1 are positioned at r=1, ... only root has max. rank
+            // -> rank = trailing zeros of id
+
+            // determine the training zeros of id
+            int rank = 0;
+            while ((id & 1) == 0) { rank++; id >>= 1; }
+            return rank;
```

Comment:
compute this using the compiler-intrinsic function System.Numerics.BitOperations.TrailingZeroCount() because it might be more efficient